### PR TITLE
Fixed failing tests on IDE-2852-build

### PR DESCRIPTION
**Pull request by szarnekow**
This PR reworks most of the scheduling of jobs, that claim to be build jobs. We had a couple of races there that caused `waitForAutoBuild` to be more something along the lines of `maybeWaitForAutoBuild`.

Issues that are addressed in this PR:
1. The build used to reset the IProject.description which caused unnecessary cache clearings and file change events. It also appeared to be redundant. This code was removed in favor of a reworked return value from the builder. It is supposed to return the projects, that the currently built project is interested in. Here we take the values directly from the manifest rather than from the dynamic project references. We do also include unresolved dependencies there, to make sure that the Eclipse builder infrastructure notifies the current project when one of its dependencies becomes newly available.
2. The build did not necessarily see changes caused by closing projects. This was caused by a separate job that claimed to be a build job, but in fact it was not. We now ask all previously interested projects whether they are still available. If not, we take their delta into account. If there is no single Xtext project available in the workspace (e.g. after a bulk close or bulk delete), we update the index in a single batch operation rather than individual jobs. This batch job offers an explicit mechanism to wait for it, which is used by the tests.
3. The external projects were built in a separate job similar to the closed project handling. It was changed along the same lines. The incremental build ensures that all pending external project builds are processed before it starts building the current project. Waiting for the pending external build is explicitly possible and used in the tests.

### CODE REVIEW

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/N4JSBuildTypeTrackingBuilder.java [hunk 1]
```diff
@@ -82,10 +155,98 @@
...
+			throw new OperationCanceledException();
+		progress.worked(2);
+		// and here
+		processedAbsentReferencedProjects(visitor);
+
+		doBuild(toBeBuilt, progress.newChild(8), BuildType.INCREMENTAL);
+	}
+
+	/*
+	 * Extract from local method to make it reusable.
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/N4JSBuildTypeTrackingBuilder.java]: comment could be changed/removed

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/instructions/BuildInstruction.java [hunk 2]
```diff
@@ -234,7 +240,11 @@ private boolean shouldGenerate(Resource resource, IProject aProject) {
 						findMaxProblemSeverity = file.findMaxProblemSeverity(null, true, IResource.DEPTH_INFINITE);
 					}
 					// the final decision to build:
-					return findMaxProblemSeverity != IMarker.SEVERITY_ERROR;
+					boolean result = findMaxProblemSeverity != IMarker.SEVERITY_ERROR;
+					if (!result) {
+						return result;
+					}
+					return result;
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/instructions/BuildInstruction.java]: the above might be refactored

File: testhelpers/org.eclipse.n4js.ui.tests.helper/src/org/eclipse/n4js/tests/builder/AbstractBuilderTest.java [hunk 3]
```diff
@@ -267,7 +280,7 @@ protected ResourceSet getResourceSet(IProject project) {
 	 */
 	protected void assertXtextIndexIsValid() {
 		// ensure no build is running while we examine the Xtext index
-		final ISchedulingRule rule = externalLibraryBuilderHelper.getRule();
+		final ISchedulingRule rule = ResourcesPlugin.getWorkspace().getRoot();
```
Comment (mmews-n4) [inline testhelpers/org.eclipse.n4js.ui.tests.helper/src/org/eclipse/n4js/tests/builder/AbstractBuilderTest.java]: why is it called "getRoot"?

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/EcmaScriptSubGenerator.java [hunk 4]
```diff
@@ -107,12 +108,14 @@ protected void internalDoGenerate(Resource resource, GeneratorOption[] options,
 
 		// if the transpile-conditions are all met, then transpile:
 		if (shouldBeCompiled(resource, monitor)) {
-
+			RaceDetectionHelper.log("About to compile %s", resource.getURI());
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/EcmaScriptSubGenerator.java]: The helper contains a guard to disable the logging (by default). I'm inclined to leave the invocations in the code. That will allow us to simply enable the logging again and see which job performs which task in what order.

File: testhelpers/org.eclipse.n4js.ui.tests.helper/src/org/eclipse/n4js/tests/builder/AbstractBuilderTest.java (hunk 3 above)
Comment (szarnekow) [reply testhelpers/org.eclipse.n4js.ui.tests.helper/src/org/eclipse/n4js/tests/builder/AbstractBuilderTest.java]: The workspace root is the logical parent of all projects in the workspace. I don't really know why they made a distinction between workspace and workspace root. Probably to decouple the workspace from the IResource concept.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/CloseProjectTaskScheduler.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	 *
+	 * Public for testing purpose.
+	 */
+	public void joinRemoveProjectJob() {
+		try {
+			// Pseudo fence to make sure that we see what we want to see.
+			synchronized (this) {
+				wait(1);
+			}
+			this.removeProjectJob.join();
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/CloseProjectTaskScheduler.java]: Tests in tight loops revealed that join does not always wait for the job to really finish. Including the wait serves as a poor attempt to ensure that the current thread really sees the correct internal state of the job instance when it tries to join.

To be honest, I do not understand why the synchronized block in the job manager is not sufficient for this, but after the synchronized block was introduced here, I do no longer see spurious races between the join and the job execution. I suspect it is related to the Java memory model, but that is only a vague feeling.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/RaceDetectionHelper.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	 *
+	 * @param messageFmt
+	 *            the message
+	 * @param messageArgs
+	 *            the arguments
+	 * @see java.util.Formatter
+	 */
+	public static void log(String messageFmt, Object... messageArgs) {
+		if (ENABLED) {
+			System.out.printf(messageFmt, messageArgs);
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/RaceDetectionHelper.java]: why not logger?

File: testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatTest.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+@Inherited
+public @interface RepeatTest {
+	/**
+	 * @return number of repetitions.
+	 */
+	int times() default 10;
+}
```
Comment (yoosiba) [inline testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatTest.java]: new line

File: testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatableTestStatement.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	}
+
+	@Override
+	public void evaluate() throws Throwable {
+		for (int i = 1; i <= times; i++) {
+			System.out.printf("Run %d of %d\n", i, times);
+			statement.evaluate();
+		}
+	}
+}
```
Comment (yoosiba) [inline testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatableTestStatement.java]: new line

File: testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatedTestRule.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+		RepeatTest repeat = description.getAnnotation(RepeatTest.class);
+		if (repeat == null) {
+			repeat = description.getTestClass().getAnnotation(RepeatTest.class);
+		}
+		if (repeat != null) {
+			return new RepeatableTestStatement(repeat.times(), statement);
+		}
+		return statement;
+	}
+}
```
Comment (yoosiba) [inline testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatedTestRule.java]: new line

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/RaceDetectionHelper.java (hunk 6 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/RaceDetectionHelper.java]: I wanted to see the message without any prefix of timestamps or anything to be able to put successful test runs and failed testruns next to each other into a diff viewer.

File: testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatTest.java (hunk 7 above)
Comment (szarnekow) [reply testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatTest.java]: Done

File: testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatableTestStatement.java (hunk 8 above)
Comment (szarnekow) [reply testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatableTestStatement.java]: Done

File: testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatedTestRule.java (hunk 9 above)
Comment (szarnekow) [reply testhelpers/org.eclipse.n4js.tests.helper/src/org/eclipse/n4js/tests/repeat/RepeatedTestRule.java]: Done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/RaceDetectionHelper.java (hunk 6 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/RaceDetectionHelper.java]: As discussed offline: We'll stick to plain sysout to have "clean" messages that can be compared.

**Pull request merged**